posts/2023-03-09-Rust-1.68.0.md

```diff
@@ -0,0 +1,132 @@
+the registry, but this has started to hit scaling limitations, with noticeable
+delays while updating that repository. The new protocol should provide a
+significant performance improvement when accessing crates.io, as it will only
+download information about the subset of crates that you actually use.
+
+To use the sparse protocol with crates.io, set the environment variable
+`CARGO_REGISTRIES_CRATES_IO_PROTOCOL=sparse`, or edit your
+[`.cargo/config.toml` file](https://doc.rust-lang.org/cargo/reference/config.html)
+to add:
+
+```toml
+[registries.crates-io]
+protocol = "sparse"
+```
+
+The sparse protocol is currently planned to become the default for crates.io in
+the 1.70.0 release in a few months. For more information, please see the prior
+[announcement](https://blog.rust-lang.org/inside-rust/2023/01/30/cargo-sparse-protocol.html)
+on the Inside Rust Blog, as well as
+[RFC 2789](https://rust-lang.github.io/rfcs/2789-sparse-index.html)
+and the current
+[documentation](https://doc.rust-lang.org/stable/cargo/reference/registry-index.html#sparse-protocol)
+in the Cargo Book.
+
+### Local `Pin` construction
+
+The new [`pin!`](https://doc.rust-lang.org/stable/std/pin/macro.pin.html) macro
+constructs a `Pin<&mut T>` from a `T` expression, anonymously captured in local
+state. This is often called stack-pinning, but that "stack" could also be the
+captured state of an `async fn` or block. This macro is similar to some crates,
+like [`tokio::pin!`](https://docs.rs/tokio/1/tokio/macro.pin.html), but the
+standard library can take advantage of `Pin` internals and [temporary lifetime
+extension](https://doc.rust-lang.org/stable/reference/destructors.html#temporary-lifetime-extension)
+for a more expression-like macro.
+
+```rust
+/// Runs a future to completion.
+fn block_on<F: Future>(future: F) -> F::Output {
+    let waker_that_unparks_thread = todo!();
+    let mut cx = Context::from_waker(&waker_that_unparks_thread);
+    // Pin the future so it can be polled.
+    let mut pinned_future = pin!(future);
+    loop {
+        match pinned_future.as_mut().poll(&mut cx) {
+            Poll::Pending => thread::park(),
+            Poll::Ready(result) => return result,
+        }
+    }
+}
+```
+
+In this example, the original `future` will be moved into a temporary local,
+referenced by the new `pinned_future` with type `Pin<&mut F>`, and that pin is
+subject to the normal borrow checker to make sure it can't outlive that local.
+
+### Default `alloc` error handler
+
+When allocation fails in Rust, APIs like `Box::new` and `Vec::push` have no way
+to indicate that failure, so some divergent execution path needs to be taken.
+When using the `std` crate, the program will print to `stderr` and abort. But
+with `#![no_std]` and the `alloc` crate, it was required to provide a
+`#[alloc_error_handler]` function, similar to `#[panic_handler]`, but that
+attribute is still unstable.
```

Comment:
It may make sense to reword that, as the attribute is likely getting removed entirely. It's currently in FCP for removal: https://github.com/rust-lang/rust/issues/51540#issuecomment-1448404145

Reply:
```suggestion
`#[alloc_error_handler]` function, similar to `#[panic_handler]`. Moving forward the plan is to treat running out of memory as normal panics.
```

Maybe something like that?

Reply:
Posted a more complete rewrite in https://github.com/rust-lang/blog.rust-lang.org/pull/1088/files#r1130197777 to incorporate this feedback, would appreciate feedback.